### PR TITLE
157094666 callback on tick

### PR DESCRIPTION
**Pull request by amaziagur**
this code review is on @wip

### CODE REVIEW

File: src/test/java/io/ticktok/support/TickPublisher.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    public static final String QUEUE_HOST = "localhost";
+    public static final String QUEUE_EXCHANGE = "exchange";
+
+    public static void publish() throws Exception {
+        Connection connection = createConnection();
+        Channel channel = connection.createChannel();
+        channel.exchangeDeclare(QUEUE_EXCHANGE, "fanout");
+
+        String message = "tick";
```
Comment (rockem) [inline src/test/java/io/ticktok/support/TickPublisher.java]: extract 3 lines

File: src/test/java/io/ticktok/support/TicktockServiceStub.java [hunk 2]
```diff
@@ -1,46 +1,55 @@
...
+    public ClockRequest lastClockRequest;
 
-        new BlinkServer(port) {{
-            post("/api/v1/clocks", (req, res) ->
-                    clockRequest = new ClockRequest(req.body(), req.header("Authorization"))
-            );
+    public TicktockServiceStub(int port, boolean validResponse) throws IOException {
+        ticktokService = new BlinkServer(port) {{
+            post("/api/v1/clocks", (req, res) -> {
+                assert req.header(AUTHORIZATION).equals(TOKEN);
```
Comment (rockem) [inline src/test/java/io/ticktok/support/TicktockServiceStub.java]: extract

File: src/test/java/io/ticktok/TicktokBrokenServerTest.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        register(() -> {});
+    }
+
+    @Test(expected = Ticktok.TicktokException.class)
+    public void failOnQueueDown() throws IOException {
+       stub =  new TicktockServiceStub(9999, false);
+        register(() -> {});
+    }
+
+    private void register(Runnable runnable) {
```
Comment (rockem) [inline src/test/java/io/ticktok/TicktokBrokenServerTest.java]: move up

File: src/main/java/io/ticktok/rest/RestTicktokClient.java [hunk 4]
```diff
@@ -10,30 +14,33 @@
...
-    public TicktokRestClient(TicktokOptions options){
+    public RestTicktokClient(TicktokOptions options){
         this.domain = options.getDomain();
         this.token = options.getToken();
         this.client = HttpClientBuilder.create().build();
     }
 
-    private void handleBodyCall(HttpEntityEnclosingRequestBase request, String entry) throws IOException {
+    private String handleBodyCall(HttpEntityEnclosingRequestBase request, String entry) throws IOException {
         request.setEntity(new StringEntity(entry));
```
Comment (rockem) [inline src/main/java/io/ticktok/rest/RestTicktokClient.java:37]: exract and move down

File: src/main/java/io/ticktok/handler/TickTokHandler.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import java.security.KeyManagementException;
+import java.security.NoSuchAlgorithmException;
+import java.util.concurrent.TimeoutException;
+
+public class TickTokHandler implements TickHandler {
+
+    @Override
+    public void handleTick(Clock clock) throws IOException, TimeoutException, NoSuchAlgorithmException, KeyManagementException, URISyntaxException {
+
+        /*
```
Comment (rockem) [inline src/main/java/io/ticktok/handler/TickTokHandler.java]: delete comments

File: src/main/java/io/ticktok/handler/TickTokHandler.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        * "exchange": "string",
+            "topic": "string",
+            "uri": "string"
+        * */
+
+        ConnectionFactory factory = new ConnectionFactory();
```
Comment (rockem) [inline src/main/java/io/ticktok/handler/TickTokHandler.java]: too long of a function

File: src/main/java/io/ticktok/handler/TickTokHandler.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+        Channel channel = connection.createChannel();
+        channel.exchangeDeclare(clock.getClockChannel().getExchange(), BuiltinExchangeType.DIRECT);
+
+        Consumer consumer = new DefaultConsumer(channel) {
+            @Override
+            public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body)
+                    throws IOException {
+                String message = new String(body, "UTF-8");
+                System.out.println(" [x] Received '" + message + "'");
```
Comment (rockem) [inline src/main/java/io/ticktok/handler/TickTokHandler.java]: no prints!

File: src/test/java/io/ticktok/TicktokBrokenServerTest.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import io.ticktok.support.TicktockServiceStub;
+import org.junit.After;
+import org.junit.Test;
+
+import java.io.IOException;
+
+import static io.ticktok.support.TicktockServiceStub.TICKTOK_SERVICE_DOMAIN;
+import static io.ticktok.support.TicktockServiceStub.TOKEN;
+
+public class TicktokBrokenServerTest {
```
Comment (rockem) [inline src/test/java/io/ticktok/TicktokBrokenServerTest.java]: to make it relate to its class I would suggest: TicktokTest_BrokenServer

File: src/test/java/io/ticktok/TicktokBrokenServerTest.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Test(expected = Ticktok.TicktokException.class)
+    public void failOnServerDown() {
+        register(() -> {});
+    }
+
+    @Test(expected = Ticktok.TicktokException.class)
+    public void failOnQueueDown() throws IOException {
+       stub =  new TicktockServiceStub(9999, false);
+        register(() -> {});
```
Comment (rockem) [inline src/test/java/io/ticktok/TicktokBrokenServerTest.java]: indent

File: src/test/java/io/ticktok/TicktokBrokenServerTest.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    private void register(Runnable runnable) {
+        new Ticktok(new TicktokOptions(TICKTOK_SERVICE_DOMAIN, TOKEN)).newClock(EVERY_5_SECONDS, runnable);
+    }
+
+    @After
+    public void tearDown(){
```
Comment (rockem) [inline src/test/java/io/ticktok/TicktokBrokenServerTest.java]: you can stop it in relevant test since it's not bother all tests

**Review by rockem — CHANGES_REQUESTED** [review]
enjoy

File: src/test/java/io/ticktok/TicktokBrokenServerTest.java (hunk 9 above)
Comment (amaziagur) [reply src/test/java/io/ticktok/TicktokBrokenServerTest.java]: Fixed

File: src/test/java/io/ticktok/TicktokBrokenServerTest.java (hunk 8 above)
Comment (amaziagur) [reply src/test/java/io/ticktok/TicktokBrokenServerTest.java]: Fixed

File: src/test/java/io/ticktok/TicktokBrokenServerTest.java (hunk 10 above)
Comment (amaziagur) [reply src/test/java/io/ticktok/TicktokBrokenServerTest.java]: FIXED

File: src/main/java/io/ticktok/handler/TickTokHandler.java (hunk 7 above)
Comment (amaziagur) [reply src/main/java/io/ticktok/handler/TickTokHandler.java]: FIXED

File: src/main/java/io/ticktok/handler/TickTokHandler.java (hunk 6 above)
Comment (amaziagur) [reply src/main/java/io/ticktok/handler/TickTokHandler.java]: FIXED

File: src/main/java/io/ticktok/handler/TickTokHandler.java (hunk 5 above)
Comment (amaziagur) [reply src/main/java/io/ticktok/handler/TickTokHandler.java]: FIXED

File: src/main/java/io/ticktok/rest/RestTicktokClient.java (hunk 4 above)
Comment (amaziagur) [reply src/main/java/io/ticktok/rest/RestTicktokClient.java:37]: not sure what to extract here

File: src/test/java/io/ticktok/TicktokBrokenServerTest.java (hunk 3 above)
Comment (amaziagur) [reply src/test/java/io/ticktok/TicktokBrokenServerTest.java]: FIXED

File: src/test/java/io/ticktok/support/TicktockServiceStub.java (hunk 2 above)
Comment (amaziagur) [reply src/test/java/io/ticktok/support/TicktockServiceStub.java]: FIXED

File: src/test/java/io/ticktok/support/TickPublisher.java (hunk 1 above)
Comment (amaziagur) [reply src/test/java/io/ticktok/support/TickPublisher.java]: not sure what to extract here

Comment (rockem) [reply src/test/java/io/ticktok/support/TickPublisher.java]: sendTick

Comment (amaziagur) [reply src/test/java/io/ticktok/support/TickPublisher.java]: FIXED

**Review by rockem — APPROVED** [review]

File: src/main/java/io/ticktok/rest/RestTicktokClient.java (hunk 4 above)
Comment (amaziagur) [reply src/main/java/io/ticktok/rest/RestTicktokClient.java:37]: FIXED

**Comment by amaziagur** [comment]
support register new clock

**Pull request merged**